### PR TITLE
samples: SMF: LVGL: SMF-based Calculator

### PR DESCRIPTION
**Pull request by glenn-andrews**
Note: This is a copy of https://github.com/zephyrproject-rtos/zephyr/pull/75555 which I managed to thoroughly wreck though a botched rebase.

This sample crates a touchscreen desk calculator based on the sample state machine in _Practical UML Statecharts in C/C++_ by Miro Samek.

Sample should build and run on any touchscreen-enabled board with sufficient resources.

Tested on `disco_l475_iot1` board with `adafruit_2_8_tft_touch_v2` touchscreen.

### CODE REVIEW

File: samples/subsys/smf/smf_calculator/smf_calculator/img/smf_calculator.png
Comment (aescolar) [inline samples/subsys/smf/smf_calculator/smf_calculator/img/smf_calculator.png:1]: Documentation and diagrams are really welcome :)
Note though, that a drawing like this is probably better suited to be stored as an svg.
(Currently this png is a 140KB binary which is not a nice thing from the repo size point of view.)

File: samples/subsys/smf/smf_calculator/src/smf_calculator_thread.c [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+}
+
+static void ready_run(void *obj)
+{
+	struct s_object *s = (struct s_object *)obj;
+
+	LOG_DBG("");
+
+	switch (s->event.event_id) {
+	case DECIMAL_POINT:
```
[the remark this answers is not in the document]
Comment (glenn-andrews) [reply samples/subsys/smf/smf_calculator/src/smf_calculator_thread.c:217]: Oops, yes. Fixed.

**Review by keith-zephyr — DISMISSED** [review]

**Review by kartben — COMMENTED** [review]
@glenn-andrews @keith-zephyr I haven't forgotten about this PR. I have been thinking about it quite a lot, actually, in the context of #67021.
What are your guys' thoughts on moving this to a better "home" that would better reflect the fact that it is effectively a comprehensive "demo" beyond just being about SMF or LVGL.

Another option is also certainly to have it under samples/subsys/smf for now, but move it if/when we have a solution to #67021.

In any case, I am hoping to go through another round of review hopefully this week.

**Comment by keith-zephyr** [comment]
> @glenn-andrews @keith-zephyr I haven't forgotten about this PR. I have been thinking about it quite a lot, actually, in the context of #67021. What are your guys' thoughts on moving this to a better "home" that would better reflect the fact that it is effectively a comprehensive "demo" beyond just being about SMF or LVGL.
> 
> Another option is also certainly to have it under samples/subsys/smf for now, but move it if/when we have a solution to #67021.
> 
> In any case, I am hoping to go through another round of review hopefully this week.

I think it makes sense to land this in `samples` now, and move the future home for Zephyr "applications/demos".

File: samples/subsys/smf/smf_calculator/src/smf_console_cmds.c [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		event.event_id = DIGIT_0;
+	} else if (c == '.') {
+		event.event_id = DECIMAL_POINT;
+	} else if ((c == '+') || (c == '-') || (c == '*') || (c == '/')) {
+		event.event_id = OPERATOR;
+	} else if (c == '=') {
+		event.event_id = EQUALS;
+	} else if ((c == 'C') || (c == 'c')) {
+		event.event_id = CANCEL_BUTTON;
+	} else if ((c == 'E') || (c == 'E')) {
```
Comment (kartben) [inline samples/subsys/smf/smf_calculator/src/smf_console_cmds.c:31]:
```suggestion
	} else if ((c == 'E') || (c == 'e')) {
```

**Review by kartben — CHANGES_REQUESTED** [review]
found one more typo, sorry!

**Review by kartben — APPROVED** [review]

**Review by keith-zephyr — APPROVED** [review]

**Pull request merged**